### ISSUE TITLE
Problem with EMongoUniqueValidator when _id is a string.

### ISSUE THREAD

**Issue by kuzdo696** [Sammaye/MongoYii#258]
Hello,

I read issue #113, and tried to override getPrimaryKey, but I still have problem. User can create a new record with custom _id, and Yii should validate that if that value was taken already, but validation returns, that $doc and $object are the same document, so page is reloading and then MongoDuplicateKeyException raise. What I should do?

PS. Sorry for my English, I'm not native English speaker (I understand what I read, I have problems with writing and tenses in English).
Regards.

**Comment by Sammaye** [comment]
Interesting it should work, can you tell me what data you are using for the _id?

**Comment by kuzdo696** [comment]
Normal strings without spaces. "test", "something", etc.

**Comment by kuzdo696** [comment]
This is my model:
```php
<?php

class Sources extends EMongoDocument {

    /**
     * @var integer
     */
    public $_id;

    /**
     * @var string
     */
    public $name;

    /**
     * @var boolean
     */
    public $active = 0;

    /**
     * @var string
     */
    public $update_date = "";

    /**
     * @var integer
     */
    public $status = 0;

    /**
     * @var subdocument
     */
    public $urls;

    /**
     * @var Sources
     */
    private $oldRecord;

    protected function afterFind() {
        $this->oldRecord = clone $this;
        return parent::afterFind();
    }

    public function collectionName() {
        return 'sources';
    }

    public static function model($className = __CLASS__) {
        return parent::model($className);
    }

    public function rules() {
        return array(
            array('_id, name', 'required', 'on' => 'new'),
            array('_id, name', 'safe', 'on' => 'new'),
            array('_id', 'EMongoUniqueValidator')
        );
    }

    public function getPrimaryKey($value = null) {
        if($value === null) {
            $value = (string) $this->{$this->primaryKey()};
        }
        return (string) $value;
    }
}
```

And action:
```php
<?php

class NewAction extends CAction {

    /**
     * @var Sources
     */
    private $source;

    public function run() {

        $request      = Yii::app()->request;
        $this->source = new Sources('new');

        if($request->getParam('ajax') === 'new-source') {
            echo CActiveForm::validate($this->source);
            Yii::app()->end();
        }

        if(is_array($request->getParam('Sources'))) {
            $this->source->attributes = $request->getParam('Sources');
            if($this->source->save()) {
                $request->redirect(Yii::app()->createUrl('sources/default'), true);
            }
        }

        $this->controller->render('new', array(
            'model' => $this->source
        ));
    }

}
```

And when I add new item with new unique _id, all is ok. But when I add second item with the same _id, CActiveForm should show me error, that this value has been taken already, but instead of it, page is reaload, so values are passed validation, and then exception is raised.

**Comment by Sammaye** [comment]
Everything looks fine with your code so I will need to dig into the source of the extension for this one

**Comment by kuzdo696** [comment]
Thanks for reply. One more note: When I don't override getPrimaryKey, page isn't reloaded because server throw exception "MongoException" with message 'Invalid object ID'.

**Comment by Sammaye** [comment]
Indeed, the default behaviour is to expect a objectID and since your return is string it throws a wobbly

**Comment by Sammaye** [comment]
I have reproduced the problem and am looking into a fix now

**Comment by Sammaye** [comment]
Ah it is a problem of using specifically the primary key for this

**Comment by kuzdo696** [comment]
So, what I should do to make it work?

**Comment by Sammaye** [comment]
I am working on a patch but it will take a little while

**Comment by Sammaye** [comment]
I have added a new unique validator, can you try it out? https://github.com/Sammaye/MongoYii/compare/8.1.7...master

**Comment by kuzdo696** [comment]
Now all it's ok :smiley:
Big thx :+1:

**Comment by Sammaye** [comment]
Awesome thanks

**Comment by Sammaye** [comment]
Released to stable 8.1.8
